File: jlpt/native_ios/Sources/Domain/Repositories/AuthRepository.swift
import Foundation

protocol AuthRepository {
    /// Broadcasts the currently persisted session, or nil when signed out. Yields once immediately
    /// with the restored-on-launch value, then again on every sign-in/sign-out transition.
    func sessionUpdates() -> AsyncStream<AuthSession?>

    func login(email: String, password: String) async -> AppResult<AuthSession>

    func register(
        email: String,
        password: String,
        displayName: String,
        targetLevel: JlptLevel?
    ) async -> AppResult<AuthSession>

    func logout() async
}

File: jlpt/native_android/app/src/main/java/com/jlpt/android/domain/repository/AuthRepository.kt
package com.jlpt.android.domain.repository

import com.jlpt.android.domain.model.AuthSession
import com.jlpt.android.domain.model.JlptLevel
import com.jlpt.android.domain.util.AppResult
import kotlinx.coroutines.flow.Flow

interface AuthRepository {
    /** Emits the currently persisted session, or null when signed out. Restored on app start. */
    val session: Flow<AuthSession?>

    suspend fun login(email: String, password: String): AppResult<AuthSession>

    suspend fun register(
        email: String,
        password: String,
        displayName: String,
        targetLevel: JlptLevel?
    ): AppResult<AuthSession>

    suspend fun logout()
}
